kickstarter/android-oss
# MBL-734: Retrofit client migrated to rxjava2 + LoginViewModel lifecycle improvements

# 📲 What

New Retrofit networking client for V1 API using RXJava2, LoginActivity + LoginViewModel adopting Jetpack ViewModel

# 🤔 Why

* Improve Lifecycle awareness
* Control over subscription dispose
* Avoid inheritance from `BaseActivity` & `ActivityViewModel`
* Stop using trello's RxLifecycle library

# 🛠 How

* Added V2 versions using RXJava2 to `environment`, in that way we can proceed witch incremental Activity/Fragment-ViewModel migration combos, and keep delivering product features not blocked by a considerable size migration. 
* Migrated `LoginActivity` & `LoginViewModel` to use Jetpack ViewModel + RxJava2, `LoginViewModel` is using the `apiClientV2` and `apolloClientV2` to execute the networking requests.
* Back transition improved for `LoginActivity` as was using deprecated methods

# 👀 See

- Login in Staging environment

https://github.com/kickstarter/android-oss/assets/4083656/4f99236c-d732-4ede-879d-355fa19efea6

- Login error handling 

https://github.com/kickstarter/android-oss/assets/4083656/33c67f0a-3323-4893-ab58-41ba62443949

- Reset password success message is handled in `LoginActivity`

https://github.com/kickstarter/android-oss/assets/4083656/a1346f92-2b42-4279-93b5-9a526ee03df3

- Back navigation before and after 

| Before 🐛 | After 🦋 |
| https://github.com/kickstarter/android-oss/assets/4083656/c175950f-45fb-4d41-8d57-81ce95aa8f06 | https://github.com/kickstarter/android-oss/assets/4083656/aff151f9-7a45-46f3-9d0d-f4709c24bebc |
|  |  |

# 📋 QA

- Log in staging
- Log in production ⚠️ 2fa is only active on Production environment so highly suggested try during QA
- Remember password success messages is tied to `onActvityResult` flow in `LoginActivity` so try that success message is shown. 
- Opening pop-up menu options on the top right menu. 

# Story 📖

[MBL-734](https://kickstarter.atlassian.net/browse/MBL-734)


[MBL-734]: https://kickstarter.atlassian.net/browse/MBL-734?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
